### PR TITLE
feat: Slack alert when VMs are on unapproved portgroups

### LINKED ISSUE #812: PCI portgroup enforcement/notification
**Is your feature request related to a problem? Please describe.**
Ensuring VMs stay on the segments they're supposed to

**Describe the solution you'd like**
I would like a new function monitoring VM portgroups.
- A VM gets tagged as PCI with vCenter tags
- A portgroup gets tagged as PCI with vCenter tags
- Any time a VM is reconfigured, the function checks the VM to see if it's marked as PCI. If so, it checks all vNICs to ensure they are on PCI segments. If any vNICs are on disallowed segments, notify via Slack.

### PR DESCRIPTION
**Pull request by kremerpatrick**
Closes #812

Signed-off-by: Patrick Kremer <pkremer@vmware.com>

## Summary

This function creates a Slack notification when VM network portgroups are out of compliance. The intent is to ensure that VMs tagged as PCI VMs remain on portgroups tagged as PCI portgroups. Any time the VM is edited and a portgroup is changed to a non-PCI portgroup, the function alerts via Slack.

## Pull Request Checklist
🚨 Please review the [guidelines for contributing](https://vmweventbroker.io/community) to this repository.

- [ ] Please ensure that you are making a pull request against the **Development** branch
- [ ] Please use the `WIP` keyword in the title of your PR if you are not ready for review
- [ ] Please ensure that you have opened a Github Issue if you are resolving/fixing a problem
- [ ] Please ensure that you have [signed](https://help.github.com/en/github/authenticating-to-github/signing-commits) all commits and that you have [squashed](https://medium.com/@slamflipstrom/a-beginners-guide-to-squashing-commits-with-git-rebase-8185cf6e62ec) all relevant commits related to your change
- [ ] Please make sure that you have tested your change locally by successfully [building and deploying the VMware Event Broker Appliance](https://vmweventbroker.io/kb/contribute-appliance) and/or [building and deploying VMware Event Router](https://vmweventbroker.io/kb/contribute-eventrouter)
- [ ] Please include any relevant screenshots and/or output as part of your testing
- [ ] Please include any documentation updates that is applicable for your changes

## Change Type

What types of changes does your code introduce to the VMware Event Broker Appliance?

_Put an `x` in all boxes that apply_

Please check the type of change your PR introduces:
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe):

## Resolved Issues

Closes #812 

## Testing Verification

Tested on VEBA 0.7.1. Tested with DVS and VSS portgroups, VMs with and without tags, portgroups with and without tags, VMs with multiple vNICs and mixed portgroup configurations

## Additional Information

* Any other details you wish to include or mention

If you have any questions/comments, feel free to reach out to team on Slack [#vcenter-event-broker-appliance](https://vmwarecode.slack.com/archives/CQLT9B5AA)

Thank you from the VEBA Team! 🥳

### CODE REVIEW

File: examples/knative/powercli/kn-pcli-pg-check/README.md [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+
+```console
+docker push ${IMAGE}
+```
+
+Update the `function_secret.json` file with the same vCenter Server credentials and configurations that you used in `test/docker-test-env-variable` and then create the kubernetes secret which can then be accessed from within the function by using the environment variable named called `FUNCTION_SECRET`.
+
+```console
+# create secret
+
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-pg-check/README.md:267]: nit: redundant newline

File: examples/knative/powercli/kn-pcli-pg-check/README.md [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+kubectl -n vmware-functions label secret pg-check-secret app=veba-ui
+```
+
+Edit the `function.yaml` file with the name of the container image from Step 1 if you made any changes. If not, the default VMware container image will suffice. By default, the function deployment will filter on the `VmReconfiguredEvent` vCenter Server Event. If you wish to change this, update the `subject` field within `function.yaml` to the desired event type.
+
+Deploy the function to the VMware Event Broker Appliance (VEBA).
+
+```console
+# deploy function
+
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-pg-check/README.md:280]: nit: redundant newline

File: examples/knative/powercli/kn-pcli-pg-check/README.md [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+Verify the container image works by executing it locally.
+
+Change into the `test` directory
+```console
+cd test
+```
+
+Update the following variable names within the `docker-test-env-variable` file.
+> Note - The sample variables are built for a vCenter function. You will need to replace them if you are authoring a function for a different purpose.
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-pg-check/README.md:64]: looks like leftover from a template?

File: examples/knative/powercli/kn-pcli-pg-check/README.md [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+> Any changes to variables must also be updated in `function_secret.yaml` and `test/docker-test-env-variable`
+
+* `VCENTER_SERVER` - IP Address or FQDN of the vCenter Server to connect to
+* `VCENTER_USERNAME` - vCenter account with permission to reconfigure distributed virtual switches
+* `VCENTER_PASSWORD` - vCenter password associated with the username
+* `VCENTER_CERTIFCATE_ACTION` - Set-PowerCLIConfiguration Action to configure when connection fails due to certificate error, default is Fail. (Possible values: Fail, Ignore or Warn)
+* `VM_WATCH_TAGS` - vCenter tags indicating a VM belongs on the PCI segment. In this example, `PCI/PCI-VM` and `PCI/PCI-VM2`.
+* `PG_WATCH_TAGS` - vCenter tags indicating a portgroup is a valid PCI segment. Int his example, `PCI/PCI-Network` and `PCI/PCI-Network2`
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-pg-check/README.md:72]:
```suggestion
* `PG_WATCH_TAGS` - vCenter tags indicating a portgroup is a valid PCI segment. In this example, `PCI/PCI-Network` and `PCI/PCI-Network2`
```

File: examples/knative/powercli/kn-pcli-pg-check/README.md [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+Update the following variable names within the `docker-test-env-variable` file.
+> Note - The sample variables are built for a vCenter function. You will need to replace them if you are authoring a function for a different purpose.
+> Any changes to variables must also be updated in `function_secret.yaml` and `test/docker-test-env-variable`
+
+* `VCENTER_SERVER` - IP Address or FQDN of the vCenter Server to connect to
+* `VCENTER_USERNAME` - vCenter account with permission to reconfigure distributed virtual switches
+* `VCENTER_PASSWORD` - vCenter password associated with the username
+* `VCENTER_CERTIFCATE_ACTION` - Set-PowerCLIConfiguration Action to configure when connection fails due to certificate error, default is Fail. (Possible values: Fail, Ignore or Warn)
+* `VM_WATCH_TAGS` - vCenter tags indicating a VM belongs on the PCI segment. In this example, `PCI/PCI-VM` and `PCI/PCI-VM2`.
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-pg-check/README.md:71]: question: are these variables then array/list types?

File: examples/knative/powercli/kn-pcli-pg-check/test/test-payload.json [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+	  "Name": "REPLACE-ME",
+	  "Vm": {
+		"Type": "VirtualMachine",
+		"Value": "REPLACE-ME"
+	  }
+	},
+	"Ds": null,
+	"Net": null,
+	"Dvs": null,
+	"FullFormattedMessage": "Reconfigured tinycore-2 on esx02.ad.patrickkremer.com in HomeLab.  \n \nModified:  \n \nconfig.hardware.device(4000).backing.port.portgroupKey: \"dvportgroup-1006\" -> \"dvportgroup-8004\"; \n\nconfig.hardware.device(4000).backing.port.portKey: \"4\" -> \"65\"; \n\nconfig.hardware.device(4000).backing.port.connectionCookie: 174899823 -> 193390024; \n\n Added:  \n \n Deleted:  \n \n",
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-pg-check/test/test-payload.json:37]: nit: example text should not contain user-specific data

File: examples/knative/powercli/kn-pcli-pg-check/README.md (hunk 1 above)
Comment (kremerpatrick) [reply examples/knative/powercli/kn-pcli-pg-check/README.md:267]: Fixed

File: examples/knative/powercli/kn-pcli-pg-check/README.md (hunk 2 above)
Comment (kremerpatrick) [reply examples/knative/powercli/kn-pcli-pg-check/README.md:280]: fixed

File: examples/knative/powercli/kn-pcli-pg-check/README.md (hunk 3 above)
Comment (kremerpatrick) [reply examples/knative/powercli/kn-pcli-pg-check/README.md:64]: Oops, deleted

File: examples/knative/powercli/kn-pcli-pg-check/README.md (hunk 4 above)
Comment (kremerpatrick) [reply examples/knative/powercli/kn-pcli-pg-check/README.md:72]: Fixed

File: examples/knative/powercli/kn-pcli-pg-check/README.md (hunk 5 above)
Comment (kremerpatrick) [reply examples/knative/powercli/kn-pcli-pg-check/README.md:71]: Yes, they then become JSON arrays

File: examples/knative/powercli/kn-pcli-pg-check/test/test-payload.json (hunk 6 above)
Comment (kremerpatrick) [reply examples/knative/powercli/kn-pcli-pg-check/test/test-payload.json:37]: Replaced

File: examples/knative/powercli/kn-pcli-pg-check/handler.ps1 [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+         }
+
+         # Check to see if the hash is empty
+         if ( $networkAdapterHash.Count -gt 0 ) {
+            $msg = "NICs using unapproved portgroups:`n"
+            Write-Host "$(Get-Date) - $($msg)"
+            # Build a list of NICs and unapproved portgroups
+            foreach ($nic in $networkAdapters) {
+               if ($networkAdapterHash.ContainsKey($nic.Id)) {
+                  Write-Host $nic.Name, $networkAdapterHash[$nic.Id]
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-pg-check/handler.ps1:255]: $(Get-Date)

File: examples/knative/powercli/kn-pcli-pg-check/handler.ps1 [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+
+            try {
+               Invoke-WebRequest -Uri $(${jsonSecrets}.SLACK_WEBHOOK_URL) -Method POST -ContentType "application/json" -Body $body
+            } catch {
+               throw "$(Get-Date) - Failed to send Slack Message: $($_)"
+            }
+
+            Write-Host "$(Get-Date) - Successfully sent Webhook ..."
+          } else {
+            Write-Host "$(Get-Date) - INFO: All NICs are on approved portgroups"
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-pg-check/handler.ps1:309]: nit: inverse logic to reduce if/else nesting

**Review by embano1 — CHANGES_REQUESTED** [review]

File: examples/knative/powercli/kn-pcli-pg-check/handler.ps1 [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+   catch {
+      throw "`nK8s secret `$env:FUNCTION_SECRET does not look to be defined"
+   }
+
+   $VM_WATCH_TAGS = ${jsonSecrets}.VM_WATCH_TAGS
+   $PG_WATCH_TAGS = ${jsonSecrets}.PG_WATCH_TAGS
+
+   $deviceChange = ($NULL -eq $cloudEventData.ConfigSpec.DeviceChange )
+   if(${env:FUNCTION_DEBUG} -eq "true") {
+      Write-Host "$(Get-Date) - DEBUG: ConfigSpec.DeviceChange Is Null? $($NULL -eq $cloudEventData.ConfigSpec.DeviceChange)`n"
```
[the remark this answers is not in the document]
Comment (kremerpatrick) [reply examples/knative/powercli/kn-pcli-pg-check/handler.ps1:83]: Fixed

File: examples/knative/powercli/kn-pcli-pg-check/handler.ps1 [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+   }
+
+   # If no devices changed, then the NIC could not have been updated to a different portgroup
+   if ($NULL -eq $cloudEventData.ConfigSpec.DeviceChange)
```
[the remark this answers is not in the document]
Comment (kremerpatrick) [reply examples/knative/powercli/kn-pcli-pg-check/handler.ps1:87]: Fixed

File: examples/knative/powercli/kn-pcli-pg-check/handler.ps1 [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+      $jsonSecrets = ${env:FUNCTION_SECRET} | ConvertFrom-Json
+   }
+   catch {
+      throw "`nK8s secret `$env:FUNCTION_SECRET does not look to be defined"
+   }
+
+   $VM_WATCH_TAGS = ${jsonSecrets}.VM_WATCH_TAGS
+   $PG_WATCH_TAGS = ${jsonSecrets}.PG_WATCH_TAGS
+
+   $deviceChange = ($NULL -eq $cloudEventData.ConfigSpec.DeviceChange )
```
[the remark this answers is not in the document]
Comment (kremerpatrick) [reply examples/knative/powercli/kn-pcli-pg-check/handler.ps1:81]: fixed

File: examples/knative/powercli/kn-pcli-pg-check/handler.ps1 [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+   if(${env:FUNCTION_DEBUG} -eq "true") {
+      Write-Host "$(Get-Date) - DEBUG: ConfigSpec.DeviceChange Is Null? $($NULL -eq $cloudEventData.ConfigSpec.DeviceChange)`n"
+   }
+
+   # If no devices changed, then the NIC could not have been updated to a different portgroup
+   if ($NULL -eq $cloudEventData.ConfigSpec.DeviceChange)
+   {
+      Write-Host "$(Get-Date) - No devices changed.`n"
```
[the remark this answers is not in the document]
Comment (kremerpatrick) [reply examples/knative/powercli/kn-pcli-pg-check/handler.ps1:89]: yes. fixed.

File: examples/knative/powercli/kn-pcli-pg-check/handler.ps1 [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+               $checkVM = $true
+               break
+            }
+         }
+
+         # Also breaking the outer loop - we don't need to compare any more VM tags if any match is found
+         if ($checkVM -eq $true) {
+            break
+         }
+
```
[the remark this answers is not in the document]
Comment (kremerpatrick) [reply examples/knative/powercli/kn-pcli-pg-check/handler.ps1:148]: Fixed

File: examples/knative/powercli/kn-pcli-pg-check/handler.ps1 [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+               if(${env:FUNCTION_DEBUG} -eq "true") {
+                  Write-Host "$(Get-Date) - DEBUG: Match found for: $($watchTag), breaking out of loop`n"
+               }
+               $checkVM = $true
+               break
+            }
+         }
+
+         # Also breaking the outer loop - we don't need to compare any more VM tags if any match is found
+         if ($checkVM -eq $true) {
```
[the remark this answers is not in the document]
Comment (kremerpatrick) [reply examples/knative/powercli/kn-pcli-pg-check/handler.ps1:145]: Yes, fixed

File: examples/knative/powercli/kn-pcli-pg-check/handler.ps1 [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+            break
+         }
+
+      }
+
+      # If the VM was tagged as a VM that needs to be monitored, perform the portgroup check here
+      if ($checkVM -eq $true) {
```
[the remark this answers is not in the document]
Comment (kremerpatrick) [reply examples/knative/powercli/kn-pcli-pg-check/handler.ps1:152]: Fixed

File: examples/knative/powercli/kn-pcli-pg-check/handler.ps1 [hunk 16]
```diff
@@ -0,0 +1,319 @@
...
+            }
+            elseif ($NULL -ne $dvPortGroupKey) {
+               if (${env:FUNCTION_DEBUG} -eq "true") {
+                  Write-Host "$(Get-Date) - DEBUG: DVS Port Group Key is $($dvPortGroupKey)`n"
+               }
+               # Unlike the VSS above, We have to build the MoRef ID for VDS
+               $vNetworkID = "DistributedVirtualPortgroup-" + $dvPortGroupKey
+            } else {
+               Write-Host "$(Get-Date) - ERROR: Could not determine network type`n"
+               throw "$(Get-Date) - ERROR: Could not determine network type for `n"
```
[the remark this answers is not in the document]
Comment (kremerpatrick) [reply examples/knative/powercli/kn-pcli-pg-check/handler.ps1:192]: fixed

File: examples/knative/powercli/kn-pcli-pg-check/handler.ps1 [hunk 17]
```diff
@@ -0,0 +1,319 @@
...
+                  $fullTag = $pgTag.Tag.Category.ToString() + "/" + $pgTag.Tag.Name.ToString()
+                  if(${env:FUNCTION_DEBUG} -eq "true") {
+                     Write-Host "$(Get-Date) - DEBUG: PortGroup Tag: $($fullTag)`n"
+                  }
+
+                  foreach ($watchTag in $PG_WATCH_TAGS) {
+                     if ($fullTag -eq $watchTag) {
+                        Write-Host "$(Get-Date) - INFO: Found a match on $($watchTag)`n"
+                        $pgMatch = $true
+                        break
```
[the remark this answers is not in the document]
Comment (kremerpatrick) [reply examples/knative/powercli/kn-pcli-pg-check/handler.ps1:235]: Fixed

File: examples/knative/powercli/kn-pcli-pg-check/handler.ps1 (hunk 7 above)
Comment (kremerpatrick) [reply examples/knative/powercli/kn-pcli-pg-check/handler.ps1:255]: Fixed

File: examples/knative/powercli/kn-pcli-pg-check/handler.ps1 (hunk 8 above)
Comment (kremerpatrick) [reply examples/knative/powercli/kn-pcli-pg-check/handler.ps1:309]: Fixed

**Review by embano1 — APPROVED** [review]
Nice work and super readable code! LGTM

**Pull request merged**